CatchupRange enum refactor: encode replay-only vs bucket-apply at type level

## Context

Follow-up from #2677. Now that Case 1b is removed, `CatchupRange` always returns `ReplayOnly` when LCL > genesis. The type still uses a single struct with an `Option<u32>` for bucket-apply checkpoint, meaning the invariant is enforced by runtime logic rather than the type system.

## Proposal

Refactor `CatchupRange` into an enum:
```rust
enum CatchupRange {
    ReplayOnly { from: u32, to: u32 },
    BucketApplyAndReplay { checkpoint: u32, replay_from: u32, replay_to: u32 },
    BucketsOnly { checkpoint: u32 },
}
```

This makes it structurally impossible to accidentally introduce a bucket-apply path when LCL > genesis without a type-level change.

## Priority

Low — the current code is correct and the INV-C15 assertion provides runtime safety. This is a "defense in depth" improvement.

## Review: Bounce-Back Cycle 1

**Reason:** CI failed (unrelated, will rebase)

**Reviewer A:** APPROVE — Verified every `calculate()` branch satisfies the new constructor assertions; the new `match` is exactly equivalent to the prior `replay_count() == 0` dispatch; tests were strengthened. ([comment](https://github.com/stellar-experimental/henyey/pull/2778#issuecomment-4466376713))
**Reviewer B:** APPROVE — `CatchupRange` has no out-of-crate callers; behavior preserved; only non-blocking note is a log-line format change (now `{:?}` Debug). ([comment](https://github.com/stellar-experimental/henyey/pull/2778#issuecomment-4466375323))
**External reviewers:** copilot-pull-request-reviewer: COMMENTED (neutral, generated 1 inline comment)
**CI:** red (3 failures, unrelated)

### CI failure analysis

Failing checks on PR #2778:

- `build / 3 test / 2 test (testing-with-pr, amd64, local, core,rpc,horizon , )` — `##[error]The action 'Run horizon ingesting test' has timed out after 4 minutes.`
- `build / 3 test / complete` — rollup of the above
- `build / complete` — rollup of the above

This is the Quickstart pipeline's `horizon ingesting test` timing out after 4 min. The same failure mode is already present on `main`:

- Quickstart run [25952505373](https://github.com/stellar-experimental/henyey/actions/runs/25952505373) on `main` (commit `3b31f4b3`) — same `Run horizon ingesting test` 4-min timeout.
- Quickstart run [25951830752](https://github.com/stellar-experimental/henyey/actions/runs/25951830752) on `main` (commit `f3a599e6`) — failure.

The PR diff is a pure type-shape refactor of `CatchupRange` inside `crates/history/` with no behavior change (both reviewers confirmed semantic equivalence). It is implausible that this diff caused a Quickstart-pipeline horizon-ingestion timeout.

### Routing

Routing back to `ready-for-doing`. `/do` Mode B should rebase on `origin/main` and re-push to retry CI. No code changes required for this bounce.

Per skill rule: if CI is still red on the same checks after rebase, the next `/review-pr` will mark this `blocked` for human investigation of the upstream Quickstart flake.

--- a/crates/history/README.md
+++ b/crates/history/README.md
@@ -48,7 +48,6 @@ flowchart LR
 | `HistoryArchiveState` | Parsed HAS document with bucket hash helpers and FutureBucket state. |
 | `CatchupManager` | End-to-end catchup orchestrator: download, verify, apply buckets, replay. |
 | `CatchupManagerBuilder` | Convenience builder for wiring archives, DB, bucket manager, and validation options. |
-| `CatchupRange` | stellar-core-compatible decision about buckets-only vs replay-only vs mixed catchup. |
 | `CatchupMode` | User-facing catchup policy: `Minimal`, `Complete`, or `Recent(n)`. |
 | `ReplayConfig` | Controls replay verification, eviction behavior, event emission, and publish backpressure. |
 | `ReplayedLedgerState` | Final replay summary derived from a ledger header and hash. |
@@ -76,25 +75,6 @@ println!("archive at ledger {}", has.current_ledger());
 println!("downloaded {} headers", headers.len());
 # Ok(())
 # }
-```
-
-```rust
-use henyey_history::{CatchupMode, CatchupRange, GENESIS_LEDGER_SEQ};
-
-let range = CatchupRange::calculate(
-    GENESIS_LEDGER_SEQ,
-    843_007,
-    CatchupMode::Recent(128),
-);
-
-assert!(range.apply_buckets());
-assert!(range.replay_ledgers());
-println!(
-    "apply at {}, replay {} ledgers from {}",
-    range.bucket_apply_ledger(),
-    range.replay_count(),
-    range.replay_first(),
-);
 ```
 
 ```rust

--- a/crates/history/src/catchup/mod.rs
+++ b/crates/history/src/catchup/mod.rs
@@ -744,92 +744,94 @@ impl CatchupManager {
 
         // Calculate the catchup range based on mode
         let range = CatchupRange::calculate(lcl, target, mode);
-        info!(
-            "Catchup range: apply_buckets={}, bucket_apply_ledger={}, replay_first={}, replay_count={}",
-            range.apply_buckets(),
-            if range.apply_buckets() { range.bucket_apply_ledger() } else { 0 },
-            range.replay_first(),
-            range.replay_count()
-        );
+        info!("Catchup range: {:?}", range);
 
-        let checkpoint_seq = if range.apply_buckets() {
-            // Apply buckets at the calculated checkpoint
-            let bucket_apply_at = range.bucket_apply_ledger();
-            info!("Applying buckets at checkpoint {}", bucket_apply_at);
+        let checkpoint_seq = match &range {
+            CatchupRange::BucketApplyAndReplay {
+                checkpoint: bucket_apply_at,
+                ..
+            }
+            | CatchupRange::BucketsOnly {
+                checkpoint: bucket_apply_at,
+            } => {
+                let bucket_apply_at = *bucket_apply_at;
+                info!("Applying buckets at checkpoint {}", bucket_apply_at);
 
-            // Download HAS
-            self.update_progress(
-                CatchupStatus::DownloadingHAS,
-                1,
-                "Downloading History Archive State",
-            );
-            let (has, has_archive_name) = self.download_and_verify_has(bucket_apply_at).await?;
+                // Download HAS
+                self.update_progress(
+                    CatchupStatus::DownloadingHAS,
+                    1,
+                    "Downloading History Archive State",
+                );
+                let (has, has_archive_name) = self.download_and_verify_has(bucket_apply_at).await?;
 
-            let scp_history = self.download_scp_history(bucket_apply_at).await?;
-            self.verify_and_persist_scp_history(&scp_history)?;
+                let scp_history = self.download_scp_history(bucket_apply_at).await?;
+                self.verify_and_persist_scp_history(&scp_history)?;
 
-            // Download buckets
-            self.update_progress(
-                CatchupStatus::DownloadingBuckets,
-                2,
-                "Downloading bucket files",
-            );
-            let bucket_hashes = self.compute_bucket_download_set(&has)?;
-            self.progress.buckets_total = bucket_hashes.len() as u32;
-            let buckets = self.download_buckets(&bucket_hashes).await?;
+                // Download buckets
+                self.update_progress(
+                    CatchupStatus::DownloadingBuckets,
+                    2,
+                    "Downloading bucket files",
+                );
+                let bucket_hashes = self.compute_bucket_download_set(&has)?;
+                self.progress.buckets_total = bucket_hashes.len() as u32;
+                let buckets = self.download_buckets(&bucket_hashes).await?;
 
-            // Apply buckets and initialize LedgerManager
-            let (checkpoint_header, checkpoint_hash) =
-                self.download_checkpoint_header(bucket_apply_at).await?;
+                // Apply buckets and initialize LedgerManager
+                let (checkpoint_header, checkpoint_hash) =
+                    self.download_checkpoint_header(bucket_apply_at).await?;
 
-            self.apply_buckets_and_init_ledger_manager(
-                &has,
-                &buckets,
-                bucket_apply_at,
-                checkpoint_header,
-                checkpoint_hash,
-                ledger_manager,
-                &has_archive_name,
-            )
-            .await?;
+                self.apply_buckets_and_init_ledger_manager(
+                    &has,
+                    &buckets,
+                    bucket_apply_at,
+                    checkpoint_header,
+                    checkpoint_hash,
+                    ledger_manager,
+                    &has_archive_name,
+                )
+                .await?;
 
-            bucket_apply_at
-        } else {
-            // No bucket application - Case 1: replay from current state.
-            // The LedgerManager is already initialized at the current LCL.
-            if !ledger_manager.is_initialized() {
-                // If we have existing state but ledger manager is not initialized,
-                // initialize it with the existing bucket state.
-                match existing_state {
-                    Some(state) => {
-                        info!(
-                            "Case 1 replay: initializing ledger manager from existing state at LCL {}",
-                            lcl
-                        );
-                        let (header, hash) = self.download_checkpoint_header(lcl).await?;
-                        ledger_manager
-                            .initialize(
-                                state.bucket_list,
-                                state.hot_archive_bucket_list,
-                                header,
-                                hash,
-                            )
-                            .map_err(|e| map_initialize_error(e, " from existing state"))?;
-                    }
-                    None => {
-                        return Err(HistoryError::CatchupFailed(
-                            "Catchup from LCL > genesis requires existing bucket lists or an initialized ledger manager"
-                                .to_string(),
-                        ));
+                bucket_apply_at
+            }
+            CatchupRange::ReplayOnly { .. } => {
+                // No bucket application - Case 1: replay from current state.
+                // The LedgerManager is already initialized at the current LCL.
+                if !ledger_manager.is_initialized() {
+                    // If we have existing state but ledger manager is not initialized,
+                    // initialize it with the existing bucket state.
+                    match existing_state {
+                        Some(state) => {
+                            info!(
+                                "Case 1 replay: initializing ledger manager from existing state at LCL {}",
+                                lcl
+                            );
+                            let (header, hash) = self.download_checkpoint_header(lcl).await?;
+                            ledger_manager
+                                .initialize(
+                                    state.bucket_list,
+                                    state.hot_archive_bucket_list,
+                                    header,
+                                    hash,
+                                )
+                                .map_err(|e| map_initialize_error(e, " from existing state"))?;
+                        }
+                        None => {
+                            return Err(HistoryError::CatchupFailed(
+                                "Catchup from LCL > genesis requires existing bucket lists or an initialized ledger manager"
+                                    .to_string(),
+                            ));
+                        }
                     }
                 }
+                lcl
             }
-            lcl
         };
 
         // Download, verify, and replay ledgers with retry.
         // Matches stellar-core's DownloadApplyTxsWork(RETRY_A_FEW).
-        if range.replay_count() == 0 {
+        if matches!(range, CatchupRange::BucketsOnly { .. }) {
             info!(
                 "Catching up to checkpoint {} (no ledgers to replay)",
                 checkpoint_seq

--- a/crates/history/src/catchup_range.rs
+++ b/crates/history/src/catchup_range.rs
@@ -153,7 +153,7 @@ impl std::fmt::Display for CatchupMode {
 ///   Case 4b checkpoint (LCL == genesis, target is a checkpoint inside the
 ///   first checkpoint period).
 #[derive(Debug, Clone, Copy, PartialEq, Eq)]
-pub enum CatchupRange {
+pub(crate) enum CatchupRange {
     /// Replay ledgers without applying buckets.
     ReplayOnly {
         /// Range of ledgers to replay (must be non-empty).
@@ -177,13 +177,13 @@ pub enum CatchupRange {
 
 impl CatchupRange {
     /// Create a range that only applies buckets (no replay).
-    pub fn buckets_only(checkpoint: u32) -> Self {
+    pub(crate) fn buckets_only(checkpoint: u32) -> Self {
         assert!(checkpoint > 0, "buckets_only checkpoint must be > 0");
         Self::BucketsOnly { checkpoint }
     }
 
     /// Create a range that only replays ledgers (no bucket application).
-    pub fn replay_only(replay: LedgerRange) -> Self {
+    pub(crate) fn replay_only(replay: LedgerRange) -> Self {
         assert!(
             replay.count > 0,
             "replay_only requires a non-empty replay range"
@@ -193,7 +193,7 @@ impl CatchupRange {
     }
 
     /// Create a range that applies buckets then replays ledgers.
-    pub fn buckets_and_replay(checkpoint: u32, replay: LedgerRange) -> Self {
+    pub(crate) fn buckets_and_replay(checkpoint: u32, replay: LedgerRange) -> Self {
         assert!(checkpoint > 0, "buckets_and_replay checkpoint must be > 0");
         assert!(
             replay.count > 0,
@@ -218,7 +218,7 @@ impl CatchupRange {
     /// # Panics
     ///
     /// Panics if preconditions are not met.
-    pub fn calculate(lcl: u32, target: u32, mode: CatchupMode) -> Self {
+    pub(crate) fn calculate(lcl: u32, target: u32, mode: CatchupMode) -> Self {
         // Validate preconditions
         assert!(
             lcl >= GENESIS_LEDGER_SEQ,
@@ -319,6 +319,7 @@ impl CatchupRange {
     }
 
     /// Get the replay range, if this variant performs replay.
+    #[cfg(test)]
     pub fn replay_range(&self) -> Option<LedgerRange> {
         match self {
             Self::ReplayOnly { replay } | Self::BucketApplyAndReplay { replay, .. } => {
@@ -329,6 +330,7 @@ impl CatchupRange {
     }
 
     /// Get the first ledger in the full range (bucket apply or replay start).
+    #[cfg(test)]
     pub fn first(&self) -> u32 {
         match self {
             Self::ReplayOnly { replay } => replay.first,
@@ -339,6 +341,7 @@ impl CatchupRange {
     }
 
     /// Get the last ledger in the full range.
+    #[cfg(test)]
     pub fn last(&self) -> u32 {
         match self {
             Self::ReplayOnly { replay } | Self::BucketApplyAndReplay { replay, .. } => {
@@ -349,6 +352,7 @@ impl CatchupRange {
     }
 
     /// Get the total count of ledgers (bucket apply counts as 1).
+    #[cfg(test)]
     pub fn count(&self) -> u32 {
         match self {
             Self::ReplayOnly { replay } => replay.count,

--- a/crates/history/src/catchup_range.rs
+++ b/crates/history/src/catchup_range.rs
@@ -138,49 +138,73 @@ impl std::fmt::Display for CatchupMode {
 
 /// Range required to perform a catchup operation.
 ///
-/// Contains information about whether to apply buckets and which ledgers to replay.
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub struct CatchupRange {
-    /// Whether to apply buckets from a checkpoint.
-    apply_buckets: bool,
-    /// Which ledger to apply buckets at (0 if !apply_buckets).
-    apply_buckets_at_ledger: u32,
-    /// Range of ledgers to replay after bucket application.
-    replay_range: LedgerRange,
+/// Three variants encode the post-#2677 invariant "bucket-apply only when
+/// LCL == genesis" at the type level:
+///
+/// - [`CatchupRange::ReplayOnly`]: pure replay of `replay.first ..=
+///   replay.last()`. Used for Case 0 (Complete from genesis), Case 1 (LCL >
+///   genesis), Case 4 fallback, and the Case 4b fallback when target is
+///   before the first checkpoint.
+/// - [`CatchupRange::BucketApplyAndReplay`]: apply buckets at `checkpoint`,
+///   then replay `replay` (which always starts at `checkpoint + 1`). Used
+///   for Case 5 (LCL == genesis, target past the first checkpoint).
+/// - [`CatchupRange::BucketsOnly`]: apply buckets at `checkpoint` and stop
+///   (no replay). Used for Case 3 (count=0 and target is a checkpoint) and
+///   Case 4b checkpoint (LCL == genesis, target is a checkpoint inside the
+///   first checkpoint period).
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum CatchupRange {
+    /// Replay ledgers without applying buckets.
+    ReplayOnly {
+        /// Range of ledgers to replay (must be non-empty).
+        replay: LedgerRange,
+    },
+    /// Apply buckets at `checkpoint`, then replay ledgers starting at
+    /// `checkpoint + 1`. Invariants: `replay.first == checkpoint + 1` and
+    /// `replay.count > 0`. Enforced by [`CatchupRange::buckets_and_replay`].
+    BucketApplyAndReplay {
+        /// Checkpoint ledger at which to apply buckets.
+        checkpoint: u32,
+        /// Range of ledgers to replay after bucket application.
+        replay: LedgerRange,
+    },
+    /// Apply buckets at `checkpoint` and stop (no replay).
+    BucketsOnly {
+        /// Checkpoint ledger at which to apply buckets.
+        checkpoint: u32,
+    },
 }
 
 impl CatchupRange {
     /// Create a range that only applies buckets (no replay).
-    pub fn buckets_only(apply_at: u32) -> Self {
-        let range = Self {
-            apply_buckets: true,
-            apply_buckets_at_ledger: apply_at,
-            replay_range: LedgerRange::empty(),
-        };
-        range.check_invariants();
-        range
+    pub fn buckets_only(checkpoint: u32) -> Self {
+        assert!(checkpoint > 0, "buckets_only checkpoint must be > 0");
+        Self::BucketsOnly { checkpoint }
     }
 
     /// Create a range that only replays ledgers (no bucket application).
-    pub fn replay_only(replay_range: LedgerRange) -> Self {
-        let range = Self {
-            apply_buckets: false,
-            apply_buckets_at_ledger: 0,
-            replay_range,
-        };
-        range.check_invariants();
-        range
+    pub fn replay_only(replay: LedgerRange) -> Self {
+        assert!(
+            replay.count > 0,
+            "replay_only requires a non-empty replay range"
+        );
+        assert!(replay.first > 0, "replay_only replay.first must be > 0");
+        Self::ReplayOnly { replay }
     }
 
     /// Create a range that applies buckets then replays ledgers.
-    pub fn buckets_and_replay(apply_at: u32, replay_range: LedgerRange) -> Self {
-        let range = Self {
-            apply_buckets: true,
-            apply_buckets_at_ledger: apply_at,
-            replay_range,
-        };
-        range.check_invariants();
-        range
+    pub fn buckets_and_replay(checkpoint: u32, replay: LedgerRange) -> Self {
+        assert!(checkpoint > 0, "buckets_and_replay checkpoint must be > 0");
+        assert!(
+            replay.count > 0,
+            "buckets_and_replay requires a non-empty replay range"
+        );
+        assert_eq!(
+            checkpoint + 1,
+            replay.first,
+            "replay must start immediately after bucket apply"
+        );
+        Self::BucketApplyAndReplay { checkpoint, replay }
     }
 
     /// Calculate the catchup range based on LCL, target, and mode.
@@ -294,101 +318,42 @@ impl CatchupRange {
         Self::buckets_and_replay(apply_buckets_at, replay)
     }
 
-    fn check_invariants(&self) {
-        // Must be applying buckets and/or replaying
-        assert!(
-            self.apply_buckets || self.replay_ledgers(),
-            "must apply buckets or replay ledgers"
-        );
-
-        if !self.apply_buckets && self.replay_ledgers() {
-            // Cases 1, 2, 4: no buckets, only replay
-            assert_eq!(self.apply_buckets_at_ledger, 0);
-            assert_ne!(self.replay_range.first, 0);
-        } else if self.apply_buckets && self.replay_ledgers() {
-            // Case 5: buckets and replay
-            assert_ne!(self.apply_buckets_at_ledger, 0);
-            assert_ne!(self.replay_range.first, 0);
-            assert_eq!(
-                self.apply_buckets_at_ledger + 1,
-                self.replay_range.first,
-                "replay must start immediately after bucket apply"
-            );
-        } else {
-            // Case 3: buckets only, no replay
-            assert!(self.apply_buckets && !self.replay_ledgers());
-            assert_eq!(self.replay_range.first, 0);
+    /// Get the replay range, if this variant performs replay.
+    pub fn replay_range(&self) -> Option<LedgerRange> {
+        match self {
+            Self::ReplayOnly { replay } | Self::BucketApplyAndReplay { replay, .. } => {
+                Some(*replay)
+            }
+            Self::BucketsOnly { .. } => None,
         }
-    }
-
-    /// Whether buckets should be applied.
-    pub fn apply_buckets(&self) -> bool {
-        self.apply_buckets
-    }
-
-    /// Get the ledger at which to apply buckets.
-    ///
-    /// # Panics
-    ///
-    /// Panics if `apply_buckets()` is false.
-    pub fn bucket_apply_ledger(&self) -> u32 {
-        assert!(
-            self.apply_buckets,
-            "bucket_apply_ledger called when apply_buckets is false"
-        );
-        self.apply_buckets_at_ledger
-    }
-
-    /// Whether ledgers should be replayed.
-    pub fn replay_ledgers(&self) -> bool {
-        self.replay_range.count > 0
-    }
-
-    /// Get the replay range.
-    pub fn replay_range(&self) -> LedgerRange {
-        self.replay_range
-    }
-
-    /// Get the first ledger to replay.
-    pub fn replay_first(&self) -> u32 {
-        self.replay_range.first
-    }
-
-    /// Get the number of ledgers to replay.
-    pub fn replay_count(&self) -> u32 {
-        self.replay_range.count
-    }
-
-    /// Get one past the last ledger to replay.
-    pub fn replay_limit(&self) -> u32 {
-        self.replay_range.limit()
     }
 
     /// Get the first ledger in the full range (bucket apply or replay start).
     pub fn first(&self) -> u32 {
-        if self.apply_buckets {
-            self.apply_buckets_at_ledger
-        } else {
-            self.replay_range.first
+        match self {
+            Self::ReplayOnly { replay } => replay.first,
+            Self::BucketApplyAndReplay { checkpoint, .. } | Self::BucketsOnly { checkpoint } => {
+                *checkpoint
+            }
         }
     }
 
     /// Get the last ledger in the full range.
     pub fn last(&self) -> u32 {
-        if self.replay_range.count > 0 {
-            self.replay_range.last()
-        } else {
-            assert!(self.apply_buckets);
-            self.apply_buckets_at_ledger
+        match self {
+            Self::ReplayOnly { replay } | Self::BucketApplyAndReplay { replay, .. } => {
+                replay.last()
+            }
+            Self::BucketsOnly { checkpoint } => *checkpoint,
         }
     }
 
     /// Get the total count of ledgers (bucket apply counts as 1).
     pub fn count(&self) -> u32 {
-        if self.apply_buckets {
-            self.replay_range.count + 1
-        } else {
-            self.replay_range.count
+        match self {
+            Self::ReplayOnly { replay } => replay.count,
+            Self::BucketApplyAndReplay { replay, .. } => replay.count + 1,
+            Self::BucketsOnly { .. } => 1,
         }
     }
 }
@@ -421,10 +386,12 @@ mod tests {
     fn test_case1_lcl_past_genesis_non_minimal() {
         // LCL is past genesis with non-Minimal mode — replay from LCL+1
         let range = CatchupRange::calculate(100, 200, CatchupMode::Complete);
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 101);
-        assert_eq!(range.replay_count(), 100);
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(101, 100)
+            }
+        );
     }
 
     #[test]
@@ -434,10 +401,12 @@ mod tests {
         // block the event loop and cause an infinite catchup loop.
         // target=127 is a checkpoint ledger, but gap=27 < threshold → replay_only.
         let range = CatchupRange::calculate(100, 127, CatchupMode::Minimal);
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 101);
-        assert_eq!(range.replay_count(), 27);
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(101, 27)
+            }
+        );
     }
 
     #[test]
@@ -445,10 +414,12 @@ mod tests {
         // Simulate the buffered-catchup scenario: LCL=61551871, gap=93.
         // Must use replay_only, not bucket download.
         let range = CatchupRange::calculate(61551871, 61551964, CatchupMode::Minimal);
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 61551872);
-        assert_eq!(range.replay_count(), 93);
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(61551872, 93)
+            }
+        );
     }
 
     #[test]
@@ -458,42 +429,45 @@ mod tests {
         // Previously this used bucket-download optimization; now it replays
         // the full gap to maintain INV-C15.
         let range = CatchupRange::calculate(61529351, 61551615, CatchupMode::Minimal);
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 61529352);
-        assert_eq!(range.replay_count(), 22264); // 61551615 - 61529351
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(61529352, 22264) // 61551615 - 61529351
+            }
+        );
     }
 
     #[test]
     fn test_case2_full_replay() {
         // count >= full replay count, do full replay
         let range = CatchupRange::calculate(1, 100, CatchupMode::Complete);
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 2);
-        assert_eq!(range.replay_count(), 99);
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(2, 99)
+            }
+        );
     }
 
     #[test]
     fn test_case3_buckets_only() {
         // count=0 and target is checkpoint, buckets only
         let range = CatchupRange::calculate(1, 127, CatchupMode::Minimal);
-        assert!(range.apply_buckets());
-        assert!(!range.replay_ledgers());
-        assert_eq!(range.bucket_apply_ledger(), 127);
+        assert_eq!(range, CatchupRange::BucketsOnly { checkpoint: 127 });
     }
 
     #[test]
     fn test_case3_minimal_non_checkpoint() {
         // count=0 but target is NOT a checkpoint
-        // This falls through to case 5
+        // This falls through to case 5: apply buckets at 63, replay 64..=100.
         let range = CatchupRange::calculate(1, 100, CatchupMode::Minimal);
-        assert!(range.apply_buckets());
-        assert!(range.replay_ledgers());
-        // Should apply buckets at checkpoint 63, replay from 64 to 100
-        assert_eq!(range.bucket_apply_ledger(), 63);
-        assert_eq!(range.replay_first(), 64);
-        assert_eq!(range.replay_count(), 37); // 100 - 63 = 37
+        assert_eq!(
+            range,
+            CatchupRange::BucketApplyAndReplay {
+                checkpoint: 63,
+                replay: LedgerRange::new(64, 37) // 100 - 63 = 37
+            }
+        );
     }
 
     #[test]
@@ -501,10 +475,12 @@ mod tests {
         // target start is in first checkpoint, full replay
         let range = CatchupRange::calculate(1, 50, CatchupMode::Recent(10));
         // target_start = 50 - 10 + 1 = 41, which is in first checkpoint (0-63)
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 2);
-        assert_eq!(range.replay_count(), 49);
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(2, 49)
+            }
+        );
     }
 
     #[test]
@@ -514,11 +490,13 @@ mod tests {
         // target_start = 200 - 50 + 1 = 151
         // first_in_checkpoint(151) = 128
         // last_before_checkpoint(151) = 127
-        assert!(range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.bucket_apply_ledger(), 127);
-        assert_eq!(range.replay_first(), 128);
-        assert_eq!(range.replay_count(), 73); // 200 - 127 = 73
+        assert_eq!(
+            range,
+            CatchupRange::BucketApplyAndReplay {
+                checkpoint: 127,
+                replay: LedgerRange::new(128, 73) // 200 - 127 = 73
+            }
+        );
     }
 
     #[test]
@@ -528,13 +506,13 @@ mod tests {
         // genesis. stellar-core handles this identically (CatchupRange.cpp
         // Case 2: count=UINT32_MAX >= full_replay_count → replay_only).
         let range = CatchupRange::calculate(1, 63, CatchupMode::Complete);
-        assert!(
-            !range.apply_buckets(),
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(2, 62)
+            },
             "Complete mode from genesis must replay, not bucket-apply"
         );
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 2);
-        assert_eq!(range.replay_count(), 62);
     }
 
     #[test]
@@ -543,12 +521,11 @@ mod tests {
         // (no replay). Replaying through protocol upgrades from genesis
         // produces different state hashes than the live network.
         let range = CatchupRange::calculate(1, 63, CatchupMode::Minimal);
-        assert!(
-            range.apply_buckets(),
+        assert_eq!(
+            range,
+            CatchupRange::BucketsOnly { checkpoint: 63 },
             "Minimal mode from genesis should use bucket-apply"
         );
-        assert!(!range.replay_ledgers());
-        assert_eq!(range.bucket_apply_ledger(), 63);
     }
 
     #[test]
@@ -556,20 +533,24 @@ mod tests {
         // Complete mode from genesis to a non-checkpoint target. No HAS is
         // available, so we must replay from genesis (fallback).
         let range = CatchupRange::calculate(1, 100, CatchupMode::Complete);
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 2);
-        assert_eq!(range.replay_count(), 99);
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(2, 99)
+            }
+        );
     }
 
     #[test]
     fn test_complete_from_non_genesis() {
         // Complete mode from a non-genesis LCL should still replay.
         let range = CatchupRange::calculate(50, 100, CatchupMode::Complete);
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 51);
-        assert_eq!(range.replay_count(), 50);
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(51, 50)
+            }
+        );
     }
 
     #[test]
@@ -578,12 +559,11 @@ mod tests {
         // The lcl > GENESIS guard on Case 2 prevents a full replay from
         // genesis; should fall through to bucket-apply.
         let range = CatchupRange::calculate(1, 63, CatchupMode::Recent(100));
-        assert!(
-            range.apply_buckets(),
+        assert_eq!(
+            range,
+            CatchupRange::BucketsOnly { checkpoint: 63 },
             "Recent with large count from genesis should bucket-apply, not replay"
         );
-        assert!(!range.replay_ledgers());
-        assert_eq!(range.bucket_apply_ledger(), 63);
     }
 
     #[test]
@@ -594,9 +574,12 @@ mod tests {
         let range = CatchupRange::calculate(1, 50, CatchupMode::Recent(40));
         // target_start = 50 - 40 + 1 = 11, first_in_checkpoint(11) = 1 <= GENESIS
         // lcl == GENESIS, target is not a checkpoint -> fallback replay
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 2);
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(2, 49)
+            }
+        );
     }
 
     #[test]
@@ -604,10 +587,12 @@ mod tests {
         // LCL > genesis, target in first checkpoint.
         // Case 1 now unconditionally replays from LCL+1 (stellar-core parity).
         let range = CatchupRange::calculate(2, 50, CatchupMode::Recent(10));
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 3);
-        assert_eq!(range.replay_count(), 48); // 50 - 2 = 48
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(3, 48) // 50 - 2 = 48
+            }
+        );
     }
 
     #[test]
@@ -619,12 +604,13 @@ mod tests {
         // target_start = 843007 - 128 + 1 = 842880
         // first_in_checkpoint(842880) = 842880 (it's at checkpoint start)
         // last_before_checkpoint(842880) = 842879
-        assert!(range.apply_buckets());
-        assert!(range.replay_ledgers());
-        // Should apply buckets at 842879, replay 128 ledgers
-        assert_eq!(range.bucket_apply_ledger(), 842879);
-        assert_eq!(range.replay_first(), 842880);
-        assert_eq!(range.replay_count(), 128); // 843007 - 842879 = 128
+        assert_eq!(
+            range,
+            CatchupRange::BucketApplyAndReplay {
+                checkpoint: 842879,
+                replay: LedgerRange::new(842880, 128) // 843007 - 842879 = 128
+            }
+        );
     }
 
     #[test]
@@ -632,13 +618,15 @@ mod tests {
         // Minimal mode from genesis to a large non-checkpoint target.
         // count=0 → target_start = 10001 (past first checkpoint).
         // Falls through to Case 5: bucket-apply at prior checkpoint, then replay.
-        let range = CatchupRange::calculate(1, 10000, CatchupMode::Minimal);
-        assert!(range.apply_buckets());
-        assert!(range.replay_ledgers());
         // checkpoint_start(10001) = 9984, apply_buckets_at = 9983
-        assert_eq!(range.bucket_apply_ledger(), 9983);
-        assert_eq!(range.replay_first(), 9984);
-        assert_eq!(range.replay_count(), 17); // 10000 - 9983 = 17
+        let range = CatchupRange::calculate(1, 10000, CatchupMode::Minimal);
+        assert_eq!(
+            range,
+            CatchupRange::BucketApplyAndReplay {
+                checkpoint: 9983,
+                replay: LedgerRange::new(9984, 17) // 10000 - 9983 = 17
+            }
+        );
     }
 
     #[test]
@@ -647,10 +635,12 @@ mod tests {
         // stellar-core parity (Case 1), LCL > genesis always replays —
         // bucket-apply optimization was removed to enforce INV-C15.
         let range = CatchupRange::calculate(50000, 70000, CatchupMode::Minimal);
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 50001);
-        assert_eq!(range.replay_count(), 20000); // 70000 - 50000
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(50001, 20000) // 70000 - 50000
+            }
+        );
     }
 
     // ── Recent(N) Case 1 regression tests ──────────────────────────────
@@ -662,10 +652,12 @@ mod tests {
         // Recent(500), lcl=100, target=10_000 → gap=9900.
         // With stellar-core parity, LCL > genesis always replays from LCL+1.
         let range = CatchupRange::calculate(100, 10_000, CatchupMode::Recent(500));
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 101);
-        assert_eq!(range.replay_count(), 9900);
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(101, 9900)
+            }
+        );
     }
 
     #[test]
@@ -673,10 +665,12 @@ mod tests {
         // Recent(10_000), lcl=100, target=5000 → gap=4900.
         // LCL > genesis → replay from lcl+1.
         let range = CatchupRange::calculate(100, 5000, CatchupMode::Recent(10_000));
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 101);
-        assert_eq!(range.replay_count(), 4900);
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(101, 4900)
+            }
+        );
     }
 
     #[test]
@@ -684,10 +678,12 @@ mod tests {
         // Recent(500), lcl=61_551_871, target=61_551_964 → gap=93.
         // LCL > genesis → replay from lcl+1.
         let range = CatchupRange::calculate(61_551_871, 61_551_964, CatchupMode::Recent(500));
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 61_551_872);
-        assert_eq!(range.replay_count(), 93);
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(61_551_872, 93)
+            }
+        );
     }
 
     #[test]
@@ -695,10 +691,12 @@ mod tests {
         // Recent(100), lcl=100, target=200 → gap=100 = N.
         // LCL > genesis → replay from lcl+1 regardless of count.
         let range = CatchupRange::calculate(100, 200, CatchupMode::Recent(100));
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 101);
-        assert_eq!(range.replay_count(), 100);
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(101, 100)
+            }
+        );
     }
 
     #[test]
@@ -707,10 +705,12 @@ mod tests {
         // With stellar-core parity, LCL > genesis always replays from LCL+1.
         // (Previously this fell through to checkpoint download — INV-C15 violation.)
         let range = CatchupRange::calculate(100, 200, CatchupMode::Recent(99));
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 101);
-        assert_eq!(range.replay_count(), 100);
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(101, 100)
+            }
+        );
     }
 
     #[test]
@@ -718,10 +718,12 @@ mod tests {
         // Recent(500), lcl=100, target=10_047 (non-checkpoint) → gap > 500.
         // With stellar-core parity, LCL > genesis always replays from LCL+1.
         let range = CatchupRange::calculate(100, 10_047, CatchupMode::Recent(500));
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 101);
-        assert_eq!(range.replay_count(), 9947); // 10047 - 100
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(101, 9947) // 10047 - 100
+            }
+        );
     }
 
     #[test]
@@ -729,10 +731,12 @@ mod tests {
         // Recent(150), lcl=2, target=200 → gap=198.
         // LCL > genesis → Case 1 returns replay_only.
         let range = CatchupRange::calculate(2, 200, CatchupMode::Recent(150));
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 3);
-        assert_eq!(range.replay_count(), 198);
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(3, 198)
+            }
+        );
     }
 
     #[test]
@@ -741,9 +745,64 @@ mod tests {
         // With stellar-core parity, LCL > genesis always replays from LCL+1.
         // (Previously this fell through to checkpoint download.)
         let range = CatchupRange::calculate(100, 200, CatchupMode::Recent(0));
-        assert!(!range.apply_buckets());
-        assert!(range.replay_ledgers());
-        assert_eq!(range.replay_first(), 101);
-        assert_eq!(range.replay_count(), 100);
+        assert_eq!(
+            range,
+            CatchupRange::ReplayOnly {
+                replay: LedgerRange::new(101, 100)
+            }
+        );
+    }
+
+    // ── Accessor / constructor invariant tests ─────────────────────────
+
+    #[test]
+    fn test_accessors_replay_only() {
+        let range = CatchupRange::ReplayOnly {
+            replay: LedgerRange::new(10, 5),
+        };
+        assert_eq!(range.first(), 10);
+        assert_eq!(range.last(), 14);
+        assert_eq!(range.count(), 5);
+        assert_eq!(range.replay_range(), Some(LedgerRange::new(10, 5)));
+    }
+
+    #[test]
+    fn test_accessors_bucket_apply_and_replay() {
+        let range = CatchupRange::BucketApplyAndReplay {
+            checkpoint: 127,
+            replay: LedgerRange::new(128, 73),
+        };
+        assert_eq!(range.first(), 127);
+        assert_eq!(range.last(), 200);
+        assert_eq!(range.count(), 74); // 73 replayed + 1 bucket apply
+        assert_eq!(range.replay_range(), Some(LedgerRange::new(128, 73)));
+    }
+
+    #[test]
+    fn test_accessors_buckets_only() {
+        let range = CatchupRange::BucketsOnly { checkpoint: 127 };
+        assert_eq!(range.first(), 127);
+        assert_eq!(range.last(), 127);
+        assert_eq!(range.count(), 1);
+        assert_eq!(range.replay_range(), None);
+    }
+
+    #[test]
+    #[should_panic(expected = "replay must start immediately after bucket apply")]
+    fn test_buckets_and_replay_rejects_gap() {
+        // replay.first must equal checkpoint + 1.
+        let _ = CatchupRange::buckets_and_replay(127, LedgerRange::new(130, 10));
+    }
+
+    #[test]
+    #[should_panic(expected = "buckets_and_replay requires a non-empty replay range")]
+    fn test_buckets_and_replay_rejects_empty_replay() {
+        let _ = CatchupRange::buckets_and_replay(127, LedgerRange::new(128, 0));
+    }
+
+    #[test]
+    #[should_panic(expected = "replay_only requires a non-empty replay range")]
+    fn test_replay_only_rejects_empty_replay() {
+        let _ = CatchupRange::replay_only(LedgerRange::empty());
     }
 }

--- a/crates/history/src/lib.rs
+++ b/crates/history/src/lib.rs
@@ -142,7 +142,7 @@ pub use catchup::{
     CatchupManager, CatchupOptions, CatchupProgress, CatchupStatus, CheckpointData,
     ExistingBucketState, LedgerData, LedgerTxData,
 };
-pub use catchup_range::{CatchupMode, CatchupRange, LedgerRange, GENESIS_LEDGER_SEQ};
+pub use catchup_range::{CatchupMode, LedgerRange, GENESIS_LEDGER_SEQ};
 pub use cdp::{
     extract_ledger_header, extract_transaction_envelopes, extract_transaction_metas, CacheStats,
     CachedCdpDataLake, CdpDataLake,

